internal/controllers/clusterextension_controller.go

```diff
@@ -202,21 +179,110 @@ func (r *ClusterExtensionReconciler) reconcile(ctx context.Context, ext *ocv1alp
 	return ctrl.Result{}, nil
 }
 
-func (r *ClusterExtensionReconciler) variables(ctx context.Context) ([]deppy.Variable, error) {
+func (r *ClusterExtensionReconciler) resolve(ctx context.Context, ext *ocv1alpha1.ClusterExtension) (*catalogmetadata.Bundle, error) {
 	allBundles, err := r.BundleProvider.Bundles(ctx)
 	if err != nil {
 		return nil, err
 	}
-	clusterExtensionList := ocv1alpha1.ClusterExtensionList{}
-	if err := r.Client.List(ctx, &clusterExtensionList); err != nil {
-		return nil, err
+
+	var successors successorsPredicateFunc = legacySemanticsSuccessorsPredicate
+	if features.OperatorControllerFeatureGate.Enabled(features.ForceSemverUpgradeConstraints) {
+		successors = semverSuccessorsPredicate
 	}
-	bundleDeploymentList := rukpakv1alpha2.BundleDeploymentList{}
-	if err := r.Client.List(ctx, &bundleDeploymentList); err != nil {
-		return nil, err
+
+	packageName := ext.Spec.PackageName
+	channelName := ext.Spec.Channel
+	versionRange := ext.Spec.Version
+
+	predicates := []catalogfilter.Predicate[catalogmetadata.Bundle]{
+		catalogfilter.WithPackageName(packageName),
+	}
+
+	if channelName != "" {
+		predicates = append(predicates, catalogfilter.InChannel(channelName))
+	}
+
+	if versionRange != "" {
+		vr, err := mmsemver.NewConstraint(versionRange)
+		if err != nil {
+			return nil, fmt.Errorf("invalid version range %q: %w", versionRange, err)
+		}
+		predicates = append(predicates, catalogfilter.InMastermindsSemverRange(vr))
+	}
+
+	// Do not include bundle versions older than currently installed unless UpgradeConstraintPolicy = 'Ignore'
+	var installedVersion string
+	if ext.Spec.UpgradeConstraintPolicy != ocv1alpha1.UpgradeConstraintPolicyIgnore {
+		bd := &rukpakv1alpha2.BundleDeployment{}
```

Comment:
I dumped a lot in here. Will split into a separate function if we decide to go with it.